docs: :building_construction: add design docs for the config file (#56)
seedcase-project/seedcase-flower: docs/design/interface/config.qmd

`template_dir=Path("templates/)`, then there must be two files in the
`templates/` directory: `templates/package.md.jinja` and
`templates/resources.md.jinja` (if outputting to `.md` files). Flower
won't check if you've included template files with different names than
what is allowed by `Section.contents`, but will restrict the names you
can use in the `Section.contents` attribute.

Theoretically, a user could create a custom style that lists all the
metadata found in the `datapackage.json` file in a single Jinja template
file, as long as it was called `package.md.jinja` (or whichever output
format they wanted).

Jinja2 template files are given only the parts of the `datapackage.json`
file that are relevant to the `Section.contents` provided. For example,
in the `contributors.md.jinja` template only the list of contributors is
made available. See the `Section` design description below for more
details.

We intentionally restrict the types of names you can use in the
`Section.contents` attribute to ensure that there is a clear mapping
between what is available in the `datapackage.json` file and what can be
output in the documentation. The available `contents` are described in

> the `Section` section below.

Reviewer: The final shape here looks nice to me!
Is `jinja_prefix` the name of the object inside the jinja template? Then `prefix` is a bit confusing to me.

> the `Section` section below.

Reviewer 2: Ah, good point. I actually had a dream about this and the design (shows how much I've been mulling over it 🤪). Yes it should be `jinja_variable` or name or object. Do you have a preference for word?

> the `Section` section below.

Reviewer 2: We can match what it says in Jinja with `jinja_variable`.

---
title: Configuration
---

This document describes the configuration settings as an interface for
both the Python classes as well as the file that contains the settings.
For the Python classes, this document includes the signature,

> docstrings, and some pseudocode as a potential implementation. It

Reviewer 3: ```suggestion
It
```
A bit repetitive from what's included in `_design-status.qmd`. I've added the pseudocode part there instead.

primarily serves as a reference for us to develop from and track what
has been finished and what remains to be done.

{{< include /docs/includes/_design-status.qmd >}}

## Overview

Given that a Data Package can be very simple with only a few data
resources or complex with many resources and detailed schemas with
hundreds of resource fields, Flower needs to be configurable to
accommodate different use cases for displaying the wide range of

> metadata. Displaying all the metadata in a single file may be sufficient

Reviewer 3: ```suggestion
metadata in `datapackage.json`. Displaying all the metadata in a single file may be sufficient
```

for a simple Data Package, but it is impractical for more complex ones.
The configuration makes it possible to customise the generated

> documentation output.

Reviewer 3: ```suggestion
documentation output accordingly.
```

Flower's configurations are meant to be used to control what type of
files are output for the human-readable documentation of the
`datapackage.json` metadata, how the contents of those files are
structured (determined by the templates), what contents are included,

> and where the files are output.

Reviewer 3: ```suggestion
Flower's configurations controls the type of
output files, the content structure of those files
(determined by the templates), what content to include,
and the output destination of those files.
```

The configuration is split into two main parts:

-   The top-level configuration that is represented by the `Config`
    class, the settings under the `[tool.seedcase-flower]` table in
    `pyproject.toml`, or the settings at the top level of the
    `_flower.toml` file. These settings dictate whether to use a
    built-in style or a custom style, along with the location of the
    template files for any custom style.
-   The lower-level configuration that is represented by one or more
    `Section` classes, the `[[tool.seedcase-flower.section]]` tables in
    `pyproject.toml`, or the `[[section]]` tables in `_flower.toml`.
    These settings dictate which contents of the `datapackage.json`
    (**not individual properties**) are included in each section of the

>     documentation and where those sections are output.

Reviewer 3: There's something about the name `Section` that to me keeps sounding like what we mean with `contents` currently.. I'll think about alternatives 🤔

template files with different names than what is allowed by
`Section.contents`, but will restrict the types of names you can use in
the `Section.contents` attribute.

Theoretically, a user could create a custom style that lists all the
metadata found in the `datapackage.json` file in a single Jinja template
file, as long as it was called `package.md.jinja` (or whichever output
format they wanted).

When making a Jinja2 template file, we designed Flower to pass the only
that parts of the `datapackage.json` file that are relevant to the
`Section.contents` provided. This makes it easier to create the
templates as the full path to the property you want to use doesn't need
to be given. For example, if we wanted to list all the contributors'
first name in the `contributors.md.jinja` template, we would only need
to use `contributors[0].given_name` (for the first contributor). See the
`Section` design description below for more details.

As will be described later, we intentionally restrict the types of names
you can use in the `Section.contents` attribute to ensure that there is
a clear mapping between what is available in the `datapackage.json` file

> and what can be output in the documentation.

Reviewer 3: ```suggestion
and what can be output in the documentation. The available `contents` are described in the `Section` section below. 
```

<!-- TODO: How might we make this more robust to changes in the Data Package standard? -->

<!-- TODO: Include a setting for Data Package version? -->

<!-- TODO: Include a prefix on the styles that includes that version? -->

::: {.callout-tip icon="false" collapse="true"}
### Pseudocode: `Config` class

```` python
from dataclasses import dataclass
from typing import Literal, Optional

# TODO: Might have to use pydantic instead to run basic checks when creating the class...?
@dataclass(frozen=True)
class Config:
    """Configuration settings for styling the metadata.

    TODO: Include link when implemented

    See the [design](LINK) for an explanation of how Config is
    designed. See the [guide](LINK) on how to set up custom styles and sections.
    See `Section` for more details on how to set up sections.

    Attributes:

>         style (Literal["quarto-book", "custom"]): Whether to use a built-in

Reviewer 2: When we implement it, yes, we can add it. This is just pseudocode. Its' the same with how I listed the content items, I don't list all of them

            sections=[
                Section(
                    output_path=Path("docs/package.qmd"),
                    contents=["package", "license", "contributors"]
                ),
                # Section that outputs each resource as a file in the `resources/` folder
                Section(
                    output_path=Path("docs/resources/"),
                    contents=["resources"]
                )
            ]
        ```
    """
    style: Literal["quarto-book", "custom"] = "quarto-book"
    template_dir: Optional[Path] = None
    sections: Optional[list[Section]] = None
````
:::

### {{< var planned >}} `Section`

The main purpose of `Section` is to allow creating separate "sections"
of the documentation, in this case, separate files to display different
parts of the `datapackage.json` metadata. There are two attributes:
`output_path` and `contents`.

The `output_path` specifies the name of the output file or folder. But
it isn't as simple as "this outputs content to this given file". That's
because some Data Package properties are arrays of objects. Sometimes it
makes sense to create a separate output file for each array item.

> When a property that we want to display is a list, like resources, we

Reviewer: You answered it in one of the comments :P

would set the output path as a folder with the special `{}` whisker
brackets in place of where the file name should be to allow creating
multiple files for each resource. For now, we only allow properties that
have a `name` sub-property to output to multiple files. The reason we
decided to use the `name` property is that it has requirements in the
Data Package specification that make it ideal for use as a file name,
specifically that it can't have spaces and must be all lowercase.

For example, creating an output file for all resources in a `resources/`

> folder would use this path `resources/{resource-name}.qmd`. Why

Reviewer: ```suggestion
For example, creating an output file for all resources in a `resources/`
folder would use the path `resources/{resource-name}.qmd`. Why
```

`resource-name` and not `name`? Because it allows for the use of created
multiple folders with custom names when multiple `name` properties exist
in the nested property. For example, the schema fields property has a

> `name` property, as does it's parent `resources` property. So to output

Reviewer: ```suggestion
`name` property, as does the resource the fields belong to. So to output
```

a file for each resource's field for each resource, we could use the

> output path

Reviewer 3: ```suggestion
`output_path`
```

`docs/resources/{resource-name}/fields/{resource-schema-field-name}.qmd`.

<!-- TODO: Not sure how this might be implemented in the Jinja templates. -->

<!-- TODO: Not sure how to do this for mult-file output. Could detect if multi-file and give only bare resource array? -->

When setting a multi-file output, the properties used within the Jinja2

> template must use `{content-id}` format. For example, if we wanted to

Reviewer 3: ```suggestion
template must use a `{content-id}` format. For example, if we wanted to
```

would only need the properties of a single resource. So to be consistent
with the input to the template, we decided on using the `{content-id}`
tag.

For the `contents` attribute, we define a set of allowed values that
correspond to selected Data Package properties we consider useful to
document in standalone pages. The allowed values and their corresponding
[JSON paths](https://en.wikipedia.org/wiki/JSONPath) to the Data Package
property are:

| Property JSON Path            | `contents` item          |
|-------------------------------|--------------------------|
| `.`                           | `package`                |
| `.contributors[*]`            | `contributors`           |
| `.licenses[*]`                | `licenses`               |
| `.resources`                  | `resources`              |
| `.resources[*].sources`       | `resource-sources`       |
| `.resources[*].licenses`      | `resource-licenses`      |
| `.resources[*].schema`        | `resource-schema`        |
| `.resources[*].schema.fields` | `resource-schema-fields` |

: The pairing between `contents` item and JSON Path of it's
corresponding Data Package property.

> ::: {.callout-tip icon="false" collapse="true"}

Reviewer 3: ```suggestion
::: {.callout-tip collapse="true" appearance="minimal"}
```
This is why I considered adding a snippet for this 😉

### Pseudocode: `Section` class

```` python
from dataclasses import dataclass
from typing import Literal, Optional
from pathlib import Path

@dataclass(frozen=True)
class Section:
    """A section of the documentation with specific `datapackage.json` properties.

    TODO: add link below.

    See the [design](LINK) for an explanation of the design of
    `Section`. See the [guide](LINK) on how to set up custom styles and sections.
    See `Config` for more details on the top-level settings.

    Attributes:

>         output_path (Optional[Path]): The output path for this section. Can be `None`

Reviewer 3: ```suggestion
        output_path (Optional[Path]): The output path for the section. Can be `None`
```

We use a configuration file to ensure that how the documentation is
built is explicitly defined and reproducible whenever someone uses the
`seedcase-flower build` command. This way, anyone who has access to the
Data Package and the configuration file can re-build the documentation
in the same way. We use TOML for the file format since there are [many
advantages to it for storing configuration
settings](https://www.anbowell.com/blog/an-in-depth-comparison-of-json-yaml-and-toml/),
especially in contrast to YAML. The two biggest advantages of TOML for
us are that you can use comments (with `#`) and that the content is
treated purely as data (whereas YAML can contain content that can be
parsed as code and executed).

The configuration file maps one-to-one to the `Config` and `Section`
classes described above. The configuration can be put either in a
separate `_flower.toml` file or in the `pyproject.toml` file under the
`[tool.seedcase-flower]` table. The ordering of which configurations are
used is as follows:

1.  If an argument is given to `style` in the `build()` CLI function,
    use that.
2.  If not, search for a `_flower.toml` file in the working directory

>     and, if found, read the configurations.

Reviewer 3: ```suggestion
    and, if found, read the configurations within that file.
```

3.  If `_flower.toml` does not exist, search for a `pyproject.toml` file
    in the working directory and, if found, read the configurations
    under the `[tool.seedcase-flower]` table.
4.  If neither file exists, use the default configurations defined in
    the `Config` class, which represent the default built-in style.

We chose to override the style configurations (which overrides any other
customisations) if given from the CLI to prevent potential configuration

> conflicts between settings in the configuration file and settings given

Reviewer: Sorry, but I still don't understand this sentence. Doesn't `style` passed on the command line override everything?

on the CLI. This way, it is clear which ordering the settings are
applied.

### {{< var planned >}} `_flower.toml`

The `_flower.toml` file contains the configuration settings for Flower.

> For a simple configuration using the built-in style, the file can be as

Reviewer 3: ```suggestion
For a simple configuration using a built-in style, the file can be as
```

simple as:

``` {.toml filename="_flower.toml"}
style = "quarto-one-page"
```

More complex customisations are where a configuration file becomes
necessary. For example:

``` toml
style = "custom"
# This would have at least the files:
# -   package.qmd.jinja
# -   contributors.qmd.jinja
# -   resources.qmd.jinja
# -   resource-schema-fields.yml.jinja
templates-dir = "path/to/templates/"

# The `index.qmd` would contain the combined `package` and `contributors` contents

> # from the Jinja2 templates.

Reviewer 3: ```suggestion
# This creates an `index.qmd` containing the combined `package` and `contributors` contents
# from the Jinja2 templates.
```

[[section]]
output-path = "docs/index.qmd"
contents = [
    "package",
    "contributors"
]

# This would create one file per resource in the `docs/resources/` folder.
# E.g. `docs/resources/demographics.qmd`, `docs/resources/blood.qmd`, etc.
[[section]]

> output-path = "docs/resources/"

Reviewer 3: From the description above, would you then need the `{}` here?

contents = ["resources"]

# This would create a YAML file for each resource with its schema fields.
# The `{resource-name}` variable would be replaced with the name of each resource.
# In this example, this would allow you to use a Quarto listing for the fields:

> # https://quarto.org/docs/websites/website-listings.html#yaml-listing-content

Reviewer 2: In each resource page. The template Jinja file would need to include the YAML header with the listing for it to work.

[[section]]
output-path = "docs/resources/{resource-name}/schema-fields.yml"
contents = ["resource-schema-fields"]
```

### {{< var planned >}} `pyproject.toml`

We allow the configuration settings to be set in the `pyproject.toml`
file under the `[tool.seedcase-flower]` table for convenience, which is
particularly useful if the Data Package is a Python project that needs
to use `pyproject.toml`. This way, no additional configuration file is
needed just for Flower.

Taking the complex example from `_flower.toml` above, the
`pyproject.toml` version would look like:

``` toml
[tool.seedcase-flower]
style = "custom"
template_dir = "templates/"

[[tool.seedcase-flower.section]]
output_path = "docs/package.qmd"
contents = [
    "package",
    "contributors"
]

[[tool.seedcase-flower.section]]
output_path = "docs/resources/"
contents = ["resources"]

[[tool.seedcase-flower.section]]
output-path = "docs/resources/{resource-name}/schema-fields.yml"
contents = ["resource-schema-fields"]
```

> :::

Reviewer 3: ```suggestion
```
Moved up

    Examples:
        ``` python
        # A section that contains only the package and contributors content types,
        # saved to the `docs/package.qmd` file.
        Section(
            output_path=Path("docs/package.qmd"),
            contents=["package", "contributors"]
       )
       ```
    """
    output_path: Optional[Path] = None
    # Needs to be one or more of the following strings.
    # TODO: Not sure if this is the best/correct way to ensure option is only one of these.
    contents: list[Literal[
        "package", "contributors", "licenses", "sources", "resources", "resource-schema",
        "resource-licenses", "resource-sources", "resource-schema-fields"
    ]]
````

## Configuration file

<!-- TODO: Replace link with our own decision "why TOML" post -->

We use a configuration file to ensure that how the documentation is
built is explicitly defined and reproducible whenever someone simply
uses the `seedcase-flower build` command. This way, anyone who has
access to the Data Package and the configuration file can re-build the
documentation in the same way. We use TOML for the file format since
there are [more advantages to it for storing configuration

> settings](https://www.anbowell.com/blog/an-in-depth-comparison-of-json-yaml-and-toml/).

Reviewer: Than yaml?

The two biggest advantages to TOML for us are that you can use comments
(with `#`) and the content is treated purely as data (e.g. YAML can
contain content that can be parsed as code and executed).

The configuration file contains a one-to-one mapping with the `Config`
and `Section` classes described above. The configuration file can either

> be named `_flower.toml` or be included in the `pyproject.toml` file

Reviewer: ```suggestion
and `Section` classes described above. The configuration can
be put either in a separate `_flower.toml` file or in the `pyproject.toml` file
```

under the `[tool.seedcase-flower]` table. The ordering of which
configurations are used is as follows:

1.  If an argument is given to `style` in the `build()` CLI function,
    use that.
2.  If not, search for a `_flower.toml` file in the working directory
    and read the configurations if it exists.
3.  If `_flower.toml` does not exist, search for a `pyproject.toml` file
    in the working directory and read the configurations under the
    `[tool.seedcase-flower]` table if it exists.
4.  If neither file exists, use the default configurations as defined in

>     the `Config` class, which represents the default built-in style.

Reviewer: ```suggestion
4.  If neither file exists, use the default configurations defined in
    the `Config` class, which represent the default built-in style.
```

own Jinja2 templates along with the custom sections where the content is
output, the `_flower.toml` file could look like:

``` toml
# This is optional if `template-dir` is given.
style = "custom"
# This would have at least the files:
# -   package.qmd.jinja
# -   contributors.qmd.jinja
# -   resources.qmd.jinja
# -   resource-schema-fields.yml.jinja
templates-dir = "path/to/templates/"

# The `index.qmd` would contain the combined `package` and `contributors` contents
# from the Jinja2 templates.
[[section]]
output-path = "docs/index.qmd"
contents = [
    "package",
    "contributors"
]

# This would create one file per resource in the `docs/resources/` folder.
# E.g. `docs/resources/demographics.qmd`, `docs/resources/blood.qmd`, etc.
[[section]]
output-path = "docs/resources/"
contents = ["resources"]

# This would create a YAML file for each resource with its schema fields.
# The `{resource-name}` variable would be replaced with the name of each resource.
# In this example, this would allow you to use a Quarto listing for the fields:
# https://quarto.org/docs/websites/website-listings.html#yaml-listing-content
[[section]]

> output-path = "docs/resources/{resource-name}/schema-fields.yml"

Reviewer: If I gave a folder as an output path, would it then create a separate file for each field?

(Oh so this is where the `{...}` notation comes in)

contents = ["resource-schema-fields"]
```

### {{< var planned >}} `pyproject.toml`

We allow the configuration settings to be set in the `pyproject.toml`
file under the `[tool.seedcase-flower]` table for convenience,

> particularly if the Data Package is also a Python project that needs to

Reviewer: ```suggestion
which is particularly useful if the Data Package is a Python project that needs to
```

use `pyproject.toml`. This way, they don't need to create an additional

> configuration file just for Flower.

Reviewer: ```suggestion
use `pyproject.toml`. This way, no additional
configuration file is needed just for Flower.
```

## {{< var wip >}} Configuration File

In addition to passing configuration options via command line flags or
Python parameters, Flower can be configured via the `pyproject.toml` or `.flower.toml` files. The
advantage of doing so is that the formatting can be easier for complex
multi-page documentation sites and that options that are frequently used
can be set once instead of included each time Flower is called from the
Terminal or Python.

### Examples

```{.toml filename="pyproject.toml"}
[tool.seedcase-flower]
style="quarto-resource-listings"
ci="quarto"
```

> Running `build` with this configuration file builds a documentation site

Reviewer 3: ```suggestion
Running `build` with this configuration file builds metadata files
```

with Quarto for the Data Package in the current working directory. The

> chosen style splits the documentation into several pages.

Reviewer 3: ```suggestion
with Quarto for the Data Package in the current working directory. Depending on the
chosen style, the metadata might be split into several pages.
```

                )
            ]
        ```
    """
    style: Literal["quarto-book", "custom"] = "quarto-book"
    template_dir: Optional[Path] = None
    sections: Optional[list[Section]] = None
````
:::

### {{< var planned >}} `Section`

The main purpose of `Section` is to allow creating separate "sections"
of the documentation, in this case, separate files to display different
parts of the `datapackage.json` metadata. There are two attributes:
`output_path` and `contents`. Each has some specific decision patterns
to them that increase their complexity but also allow for some powerful
flexibility.

For the `output_path`, this dictates what the name of the file or output
folder should be. But it isn't as simple as "this outputs content to
this given file". That's because specific properties of the
`datapackage.json` file are arrays (or Python lists). For instance, you
can have more than one data resource, which itself has many properties
that also could have many of them such as having many fields in its
schema. So we need some way to output a list of data resources to a
corresponding list of output files. That's where some complexity comes

> in.

Reviewer: Maybe shorten this a bit?

E.g. something in the spirit of:
Some Data Package properties are arrays of objects. Sometimes it makes sense to create a separate output file for each array item.

When a property that we want to display is a list, like resources, we
can set an output path as a folder with the special `{}` whisker
brackets in place of where the file name should be to allow creating

> multiple files for each resource. Since we need some way to replace the

Reviewer: Maybe a bit too much implementation detail, or maybe I'm misunderstanding

reference a resource's properties with the JSON path syntax of
`.resources[{resource-id}].title` (for the `title` property of each
resource). Then Flower will loop through each resource property and
insert the number for that item in the resource. In this case, if there
were two resources, Flower would resolve it to `.resources[0].title` in
one output template file and `.resources[1].title` in another output
template file.

For the `contents` attribute, we decided on using a specific set based
on whether a property in the `datapackage.json` is a list or whether it
represents a distinct "entity" that might be worthy of having a
dedicated page of documentation. Based on that, the properties as JSON
paths are `.contributors`, `.licenses`, `.sources`, `.resources`,
`.resources[*].schema`, `.resources[*].sources`,
`.resources[*].licenses`, and `.resources[*].schema.fields`. But there's
also the top level `.` (all the Data Package properties). So simplifying
those, the `contents` we allow are `package` (for the `.` top-level),
`contributors`, `licenses`, `sources`, `resources`, and for individual
resource properties `resource-schema`, `resource-licenses`,

> `resource-sources`, and `resource-schema-fields`. So semantically, the

Reviewer: ```suggestion
For the `contents` attribute, we define a set of allowed values that correspond
to selected Data Package properties we consider useful to document in
standalone pages. The allowed values and their corresponding JSON paths
are: <list them as pairs>
So semantically, the
```

```` python
from dataclasses import dataclass
from typing import Literal, Optional
from pathlib import Path

@dataclass(frozen=True)
class Section:
    """A section of the documentation with specific `datapackage.json` properties.

    TODO: add link below.

    See the [design](LINK) for an explanation on the decisions for the design of
    `Section`. See the [guide](LINK) on how to set up custom styles and sections.
    See `Config` for more details on the top-level settings.

    Attributes:
        output_path (Optional[Path]): The output path for this section. Can be `None`
            when a style outputs to the terminal such as when using `view()`.
            If a directory is provided, files will be created for each content type that has
            a `name` property (e.g. `resources` or `resource-schema-fields`) For example,
            if `output_path` is `Path("docs/")` and `contents` is `["resources"]`,
            then each resource will be output to `docs/{resource_name}.md` (or whichever output
            format is used in the Jinja2 template files). If a file path is provided, all contents
            within the `Section` will be output to that single file.

>         contents (list[Literal[...]]): List of content to include in this

Reviewer: ```suggestion
        contents (list[Literal[...]]): List of content items to include in this
```

> ## {{< var wip >}} Configuration File

Author: I have some reservations around the need of a config file for flower, which I split out into https://github.com/seedcase-project/seedcase-flower/issues/57 to be discussed there.

Reviewer 2: A config file is an important part of reproducibly building the documentation. Without it, users would need to list or put somewhere what the command they used was. Considering most projects don't use some type of build pipeline tool like `justfile`, and if they do, they don't always put dev type work in there like we do. How would you suggest ensuring explicit instructions and a reproducible build without the use of a config file. I can't imagine typing out `flower build --contents ... --section ... blah blah ...` as one super long line to be a very ergonomic approach. Compared to putting the config settings in a file, editing via an editor, and no matter what, if someone gets the project running `flower build` would ensure consistent doc generation.

In addition to passing configuration options via command line flags or

> python parameters, `flower` can be configured via `pyproject.toml`. The

Author: I went with just pyproject.toml here for simplicity and so that users don't have to spend time considering which config file to use. I don't see an immediate benefit in supporting a separate .flower.toml file now that pyproject nicely allows all python related config to be read from one place. That seems beneficial over having multiple different config files scattered in the root dir.

We can decide and standardize this across proejcts, e.g. https://check-datapackage.seedcase-project.org/docs/design/interface#configuration-file

Author: I don't think it is a difficult coding problem to implement, just that it adds unnecessary complexity to the interface. It's not that I'm strongly against it, I just don't understand what value is added by two possible names/location for the config file instead of one. But I don't feel strongly about this so feel free to change to what you prefer.

advantage of doing so is that the formatting can be easier for complex
multi-page documentation sites and that options that are frequently used

> can be set once instead of included each time flower is called from the

Reviewer 2: ```suggestion
can be set once instead of included each time Flower is called from the
```

command line or Python.

### Examples

```{.toml filename="pyproject.toml"}
[tool.flower]
style="quarto-resource-listings"

> ci="quarto"

Reviewer 2: What is this? `ci` wasn't in the brainstorming issue.

Author: Yes, it was. You suggested it here https://github.com/seedcase-project/seedcase-flower/issues/5#issuecomment-3617319683 and @signekb agreed in https://github.com/seedcase-project/seedcase-flower/issues/5#issuecomment-3637196157

```

> Running `build` with this configuration file builds a documentation site

Author: What happens if I am in a directory containing a config file and choose to view/build a url or a file in a different location; should the config in CWD be read or the config file at PATH? What if only one exists?

Reviewer 2: This is why I think if an option is given in the CLI that no config file is read. It's either CLI args or the config file, not both. There might be some weird interactions that we can't foresee if they do some type of merging going on, which from an implementation point of view is not easy.

in qmd format for the data package in the current working directory. The
chosen style splits the documentation into several pages. Template files
are included for setting up continuous integration with quarto on

> GitHub.

Reviewer 2: ```suggestion
chosen style splits the documentation into several pages.
```

I think we should keep the focus of Flower to be razor sharp on only taking the `datapackage.json` and generating files from that. Nothing else. Keep it simple and do one thing really well. :relaxed: The `template-data-package` can include more options, which is where other files can be incorporated. Not here.

Author: I'm ok to remove this if you no longer want it, but see https://github.com/seedcase-project/seedcase-flower/pull/56/changes#r2680472954 for why I included it

---

```{.toml filename="pyproject.toml"}
[tool.flower]
style="custom"

[[tool.flower.section]]

> template="1st-page.html"

Reviewer 2: This doesn't match at all what we have in the brainstorming issue. Can you match what we have there? Otherwise, if you have other ideas, put them there, not here.

Author: I understood the brainstorming as a starting point rather than a final spec and that we can continue the discussion/refinement in the PRs since we will have new thoughts (and examples) as we try to write the documentation for the features we have brainstormed.

I meant to write a note to call attention to this deviation though and include some reasoning, but I missed that when I made that PR. Sorry about that, I added a note in https://github.com/seedcase-project/seedcase-flower/pull/49/changes#r2681267564 now where we can discuss. I think it is easier to discuss in separate threads in the PR when there are details about multiple different parts of the interface; it gets messy quickly trying to quote reply to long comments in the issues and we can resolve threads in the PR indicating what is done.

content="package"

[[tool.flower.section]]
template="2nd-page.html"
content=[
    "contributors",
    "licenses",
]
```

Running `build` with this configuration file builds a documentation site
in html format for the data package in the current working directory.
The custom style of the documentation creates one page with the package
metadata and a second page with info on the contributors and license.

> Both pages are created using a custom HTML template.

Reviewer 2: Yea, I think with the latest comments in the brainstorming issue you can ignore that comment. The sections and content and templates work better than the include exclude ideas

    See the [design](LINK) for an explanation of how Config is
    designed. See the [guide](LINK) on how to set up custom styles and sections.
    See `Section` and `Content` help for more details on how to set up the sections.

    Attributes:
        style (Style): Whether to use a built-in style or a custom one.  Default is

>             the `Style.quarto_one_page` style. If you want to use a custom style, set

Reviewer 3: ```suggestion
           `Style.quarto_one_page`. If you want to use a custom style, set

```

4.  If neither file exists, use the default configurations defined in
    the `Config` class, which represent the default built-in style.

To provide a clear precedence for applying settings, we decided
that styles specified on the CLI override settings in the configuration
file. Though, using `"custom"` style on the CLI won't do anything, as
the custom style requires a path to the template directory as well
as the different sections for the content, which can't be provided on

> the CLI.

Reviewer 2: Yea maybe. Or we just don't allow custom to be used on the CLI :thinking:

> the CLI.

Reviewer: Hmm yeah, I guess not allowing it is cleaner maybe
